docs/index.bs

```diff
@@ -1545,6 +1552,65 @@ spec: storage; urlPrefix: https://storage.spec.whatwg.org/
   </section>
 
+  <section>
+    <h3 id="installevent-interface">{{InstallEvent}}</h3>
+
+    <pre class="idl">
+      [Exposed=ServiceWorker]
+      interface InstallEvent : ExtendableEvent {
+        Promise&lt;undefined&gt; registerRouter((RouterRule or sequence&lt;RouterRule&gt;) rules);
+      };
+
+      dictionary RouterRule {
+        required RouterCondition condition;
+        required RouterSourceEnum source;
+      };
+
+      dictionary RouterCondition {
+        USVString urlPattern;
+      };
+
+      enum RouterSourceEnum { "network" };
+    </pre>
+
+    <section>
+      <h4 id="register-router-method">{{InstallEvent/registerRouter(rules)|event.registerRouter(rules)}}</h4>
+
+      {{InstallEvent/registerRouter(rules)}} registers this service worker the rules to offload simple tasks that the fetch handler does.
+
+      <dfn method for="InstallEvent"><code>registerRouter(|rules|)</code></dfn> method *must* run these steps:
+
+        1. Let |routerRules| be a list of {{RouterRule}} dictionaries.
+        1. If |rules| is a {{RouterRule}} dictionary:
+            1. Run [=VerifyRouterRule=] algorithm with |rules|.
+            1. Append |rules| to |routerRules|.
+        1. If |rules| is a sequence of {{RouterRule}} dictionaries.
+            1.  for each |rule| in |rules|:
+                1. If running [=VerifyRouterRule=] algorithm with |rule| returns false, <a>throw</a> a <code>TypeError</code>.
+                1. Append |rule| to |routerRules|.
+        1. Set |routerRules| to [=/service worker=]'s [=static router rules object=].
+
+      <section>
+        <h5 id="verify-router-rule-algorithm"><dfn>VerifyRouterRule</dfn></h5>
+          : Input
+          :: |rule|, RouterRule
+          : Output
+          :: a boolean
+
+          1. If |rule|'s |condition| is empty, and abort these steps.
+          1. If |rule|'s |condition|'s |urlPattern| is empty, return false..
+          1. If |urlPattern| contains a regular expression, return false.
+
+             Note: Since running a user-defined regular expression has a security concern, it is prohibited.
+
+          1. If |urlPattern| cannot be [=urlpattern/parsed=], return false.
```

Comment:
This is not appropriately rigorous. Given a string (e.g. `|rule|["{{RouterRule/condition}}"]["{{RouterCondition/urlPattern}}"]`) we can then feed it to the [parse a pattern string](https://wicg.github.io/urlpattern/#parse-a-pattern-string) algorithm, or maybe the [parse a constructor string](https://wicg.github.io/urlpattern/#parse-a-constructor-string) algorithm. Which one do you want?

You then will get a return value from that. What kind of return value do you consider "cannot be parsed"?

Finally, you probably will want to store the result of parsing, so that you can execute the pattern later.

Reply:
It might be the "parse a constructor string" algorithm in the future, but it should currently be the "parse a pattern string" algorithm, and only supports pathname.  I clarified that.
I clarified what "cannot be parsed" mean here.  TypeError by parser is "cannot be parsed".
As URLPattern spec do, let me make urlPattern has an associated component.  I am not sure what is an appropriate way to write it down.